### ISSUE TITLE
Invalid recursive values should be a compiler error, not a runtime exception

### ISSUE THREAD

**Issue by evancz** [elm-lang/elm-compiler#873]
Right now, Elm will not yell at you for writing invalid recursive values like `ones = 1 :: ones` where a value is defined directly in terms of itself. In a strict language like Elm, this means you'd spend an infinite amount of time building the entire structure!

### Current Behavior

Right now, invalid code like this leads to a runtime exception. In some ways that is better than infinite looping forever, and these exceptions happen very reliably. It is not conditional on anything, you just need to evaluate the code. So it's not *awful*, but it is also something that beginners run into a decent amount.

### Solution

Only permit recursion when there is at least one lambda between the definition and the usage. So `x = x` is not allowed, but `fact n = if n <= 0 then 1 else n * fact (n - 1)` is allowed because the definition implicitly introduces a lambda.

### Timeline

This bug is kind of tricky to resolve, but it is not crazy. From a raw time efficiency standpoint, it is probably best to batch with work on Dead Code Elimination. I find it super annoying when people say "but Elm *can* have runtime exceptions" as if a compiler bug with a clear resolution is the same as runtime exceptions in JavaScript, so I'd like to do it sooner than DCE if possible.

### Other Info

  - Check out https://github.com/elm-lang/core/issues/305 which shows how the *order* of definitions can be important in generated code. Probably should be done together.
  - Xavier has some very useful analysis [here](https://groups.google.com/forum/m/#!msg/comp.lang.ml/4eK7ucslWRw/CDc9S5AGYSYJ) about how it works in OCaml.

**Comment by mitchmindtree** [comment]
I think this is the issue I'm running into, but just thought I'd mention it here in case - here's the runtime error:

```
Uncaught TypeError: Cannot read property 'x' of undefined       main.elm?debug:4887

Uncaught Error: The notify function has been called synchronously!
This can lead to frames being dropped.
Definitely report this to <https://github.com/elm-lang/Elm/issues>    main.elm?debug:4887
    notify                  debug.js:13442
    notifyWrapper           main.elm?debug:6243
    (anonymous function)         debug.js:13466
    (anonymous function)          main.elm?debug:4887

Uncaught Error: The notify function has been called synchronously!
This can lead to frames being dropped.
Definitely report this to <https://github.com/elm-lang/Elm/issues>
```

I'm quite sure it's related to a large anonymous recursive function where I'm detecting collisions for a large number of edges, perhaps a little big to post here.

I'm still new to purely functional languages and i'm not sure I follow Xavier's comment very well - is there anything I should do / look out for in particular to avoid hitting this in the meantime?

**Comment by evancz** [comment]
I think a good way to figure it out would be to describe the problem on [the mailing list](https://groups.google.com/forum/#!forum/elm-discuss) along with the actual Elm code in question. If you have a record defined in terms of itself, that would cause this issue. As far as I know, you can get around this, but I'd have to see the code to give more specific advice.

**Comment by mitchmindtree** [comment]
Ahh thanks for the reply Evan! I don't believe I have any records defined in terms of themselves, though here's the function that contains the lengthy recursive function I mentioned (it's not quite as big as I thought after removing all of my `Debug.watch`s).

```Elm
 -- Collision between two polygons with a relative velocity.                                     
 polygon : (Shape, Position) -> (Shape, Position) -> Vector -> Maybe Collision                   
 polygon (shapeA, posA) (shapeB, posB) velocity =                                                
     let pointsA = Shape.points shapeA posA                                                      
         pointsB = Shape.points shapeB posB                                                      
         edges = Shape.sides shapeA posA ++ Shape.sides shapeB posB                              
         -- Recursive function for iterating through all edges.                                  
         f edges minIntervalDistance translationAxis current will =                              
             case edges of                                                                       
                 [] -> (minIntervalDistance, translationAxis, current, will)                     
                 edge :: edges ->                                                                
                                                                                                 
                     let -- Find if the polygons are currently intersecting.                     
                         axis = Line.perpendicular edge |> Vector.unit                           
                         -- Find the projection of both A and B on the perpendicular axis.       
                         (minA, maxA) = projectPolygon axis pointsA                              
                         b            = projectPolygon axis pointsB                                                            
                                                                                                 
                         -- Check if A and B are currently intersecting.                         
                         newCurrent = if intervalDistance (minA, maxA) b > 0                     
                                      then False else current                                    
                                                                                                 
                         -- Now find if the polygons *will* intersect.                           
                         velocityProjection = Vector.dot axis velocity                           
                         (min, max) = if velocityProjection < 0                                  
                                         then (minA + velocityProjection, maxA)                  
                                         else (minA, maxA + velocityProjection)                  
                         -- Check if the velocity projection of A and B are intersecting.        
                         intervalDistance' = intervalDistance (min, max) b                       
                         newWill = if intervalDistance' > 0 then False else will                 
                                                                                                 
                         -- If the polygons are not intersecting and won't intersect, exit.      
                         -- Otherwise continue.                                                  
                     in  if not newCurrent && not newWill                                        
                             then (minIntervalDistance, translationAxis, newCurrent, newWill)    
                             else let absIntervalDistance = abs intervalDistance'                
                                      -- Check if the new intervalDistance is the shortest.      
                                      (dist, axis) =                                             
                                          if absIntervalDistance < minIntervalDistance           
                                             then let d = Vector.sub posA posB |> Vector.dot axis
                                                      axis = if d < 0 then Vector.neg axis       
                                                                      else axis                  
                                                  in  (absIntervalDistance, axis)                
                                             else (minIntervalDistance, translationAxis)         
                                      -- Then continue through the list.                         
                                  in  f edges dist axis newCurrent newWill                       
                                                                                                 
         -- Recurse through edges to find the minimum interval distance, the translation axis
         -- and whether or not the polygons are intersecting or will intersect at their current velocities.                                                   
         (minIntervalDistance, translationAxis, current, will) =                                 
             f edges 1000000 {x=0,y=0} True True     
         -- The minimum translation vector can be used to push the polygons apart.         
         translationVector = Vector.mulScalar translationAxis minIntervalDistance                            
                                                                                                 
                        
     in  if will                                                                                 
             then translationVector |> Will |> Just                                              
             else if current then Just Current else Nothing                                      
```

If you can't see anything strange at a glance, no problem! I may approach the google group instead :+1:

**Comment by delitescere** [comment]
FYI http://stygianguest.github.io/Sierpinski/PerformancePlot.html fires this exception when the simulation is completing.

**Comment by thSoft** [comment]
@delitescere Are you sure? At me, it fails with `Error: Function 'tail' expects a non-empty list!` instead.

**Comment by delitescere** [comment]
@thSoft Yes, quite sure. I get both errors in the browser console on Chrome and Safari (on OS/X).

Perhaps the NEList error is leading to the notify function being called synchronously (however, correlation != causation).

**Comment by rtfeldman** [comment]
In the near term, it would be great if `let ones = 1 :: ones` were a compiler error.

That would mean one less way for things to crash in production, and it could always be relaxed later to allow for implementations that can work.

**Comment by mgold** [comment]
`x = x` is another one that comes up frequently.

**Comment by seanhess** [comment]
This surprised me today:

    let res = case model.form of
            Nothing -> res
            Just form -> Session.applyDefaults form model.session.userInfo responses

Where I meant to say

              Nothing -> model.responses

It's a easy mistake to make. I definitely would have preferred to have the compiler yell at me. 

Also, it didn't hang or anything, it just crashed with `Cannot read property 'ctor' of undefined`, which was cryptic. See here for more info: https://github.com/elm-lang/core/issues/478

**Comment by dan-tull** [comment]
I hit an issue today and when I followed the trail of issue breadcrumbs, I was led here. Boiled down to simplest form it looks like this:

```foo = { bar = True }
test a b =
  a
res = test foo res.bar
```

Which results in the error "undefined is not an object (evaluating 'res.bar')" which was fairly obvious in the more complex case I originally hit, but feels like it should be a compiler instead of a runtime error.

If there's only one parameter to test, there is a compiler error about an infinite/self-referential type. Some other similar variations also result in compiler errors, but this one hits a runtime problem, so I thought I'd included it in the thread for reference.

**Comment by markhamburg** [comment]
A shorter version of Dan's error: `res = (,) "foo" ( let ( x, y ) = res in x )`

**Comment by jvoigtlaender** [comment]
I think an *extra* warning is due about this issue here. So far, all the examples given were such that due to the non-proper dealing with recursively defined values one could get runtime errors. Bad, but not as bad as getting return values that are wrong. Unfortunately, in the presence of the tail recursion optimization in the compiler, such cases can presently happen as well. Consider this program:
```elm
import Graphics.Element

f n =
  if n > 0 then
    let
      xs =
        [1,2,3]
    in
      f 0
  else
    let
      xs =
        0 :: xs
    in
      List.take 3 xs

main =
  Graphics.Element.show (f 1)
```
Due to the `xs = 0 :: xs`, there are two things one could reasonably expect to happen here:

1. `[0,0,0]` is returned, because this `xs` is interpreted as an infinite list of zeros
2. a runtime error occurs, because this `xs` is undefined

But what actually happens right now is that `[0,1,2]` is returned, completely outside the semantics of Elm. It has to do with tail recursion, as one can see by replacing the `f 0` call by `[] ++ f 0`.

**Comment by lydell** [comment]
Is the runtime error caused by the following program related to this issue?

```elm
module Main (..) where

import Signal
import Task
import Effects
import Html exposing (text)
import StartApp


main =
  app.html


port tasks : Signal (Task.Task Effects.Never ())
port tasks =
  app.tasks


app =
  StartApp.start
    { init = ( {}, Effects.none )
    , view = (\address model -> text "Hello, world!")
    , update = (\action model -> ( model, Effects.none ))
    , inputs = [ Signal.map identity app.model ]
    }
```

Note how `app` is used inside the definition of `app`.

**Comment by jvoigtlaender** [comment]
@lydell, yes. That compiles to:
```js
   var app = $StartApp.start({init: {ctor: "_Tuple2"
                                    ,_0: {}
                                    ,_1: $Effects.none}
                             ,view: F2(function (address,model) {
                                return $Html.text("Hello, world!");
                             })
                             ,update: F2(function (action,model) {
                                return {ctor: "_Tuple2",_0: model,_1: $Effects.none};
                             })
                             ,inputs: _U.list([A2($Signal.map,
                             $Basics.identity,
                             app.model)])});
```
The problem there is obvious.

**Comment by mgold** [comment]
After combing through the many duplicates, here's a summary of the cases that should be forbidden, but currently aren't (and typically blow up at runtime).

```elm
-- a constant assigned to itself (constant = non-function value)
bad = bad

-- a constant assigned to a function application in which it is an argument
bad = identity bad
bad = bad + 7
-- or recursively
bad = 2 * (bad + 1)
-- but allow variable shadowing
bad = let bad = 2 in bad -- OK

-- a constant assigned to a function application that takes a lambda where the constant is used
bad = List.map (\_ -> List.length bad) [0]
-- or recursively
bad = 3 :: List.map (\_ -> List.length bad) [0]

-- a record extended with itself
bad = { bad | field = "invalid" }

-- mutually recursive constants (rejected by repl but accepted in a file)
a = b
b = a
-- mutually recursive type aliases are handled well; use as inspiration for constants?
```
If there was any precedent for PRs getting merged, I've give it a shot.

**Comment by jvoigtlaender** [comment]
Max, with all respect, but given what the above post implies about how you think about this and what your likely approach would be, I'd advise you to not start writing that PR. You seem to think about this too much as a simple structural syntactic test. For example in the "a constant assigned to a function application that takes a lambda where the constant is used" case. You shouldn't really prevent somebody from writing this code:
```elm
bad = List.filter (\n -> if n > 0 then List.length bad == 2 else True) [0]
```
But you sure are headed down the road of preventing that code, or then trying to special case `if-then-else` to counter my objection, etc. It's not going to work that way.

**Comment by mgold** [comment]
I'm not going to write the PR because it won't get merged and would be wasted effort. My main goal was to write down the exact semantics of what should be disallowed, so that someone else (probably Evan) would have an easier time fixing the issue. I also hoped to summarize tedious conversations, not create more of them.

Because I wanted those semantics to be as simple as possible, it's true that I exclude valid definitions such as yours. But: what  applied not to `[0]` but a list from an incoming port? Shouldn't we prevent users from writing that code? Preventing that code but allowing your example seems very complex for no real benefit. Your example isn't something that someone would intentionally, usefully write; it's a contrived edge case. If you're going to hide something awful in an if-branch, it should be `Debug.crash`, not `TypeError: Cannot read property 'ctor' of undefined` (replace `[0]` with `[1]`). I personally am willing to disallow it in favor of fixing the bug.

Would you please justify why you think a  "simple structural syntactic test" is "not going to work that way"?

**Comment by jvoigtlaender** [comment]
> Would you please justify why you think a "simple structural syntactic test" is "not going to work that way"?

Because my answer to this:
> Shouldn't we prevent users from writing that code?

is "No". The code I gave as example should not be prevented, also not if it looks like this:
```elm
bad = List.filter (\n -> if n > 0 then List.length bad == 2 else True) xs
```
with `xs` coming from wherever, a port, a top-level definition, or actually being some arbitrary complex expression, as in:
```elm
fun arguments =
  let
    bad = List.filter (\n -> if n > 0 then List.length bad == 2 else True) (some expression using the arguments)
  in
    ... whatever ...
```
I was *not* saying that the reason I wanted to not prevent that definition is that for the specific list element `0` in `[0]` we know something about which way the filter predicate will go. That's not my point.

**Comment by jvoigtlaender** [comment]
Maybe a more constructive answer to:
> Would you please justify why you think a "simple structural syntactic test" is "not going to work that way"?

Because it would *also* prevent the currently legitimate ways of writing recursive Json decoders using `lazy` from http://package.elm-lang.org/packages/elm-lang/lazy/1.0.0/Lazy, for example.

**Comment by mgold** [comment]
> I was not saying that the reason I wanted to not prevent that definition is that for the specific list element

Okay, I misunderstood you.

Thank you for pointing out recursive JSON decoders; those certainly are *not* contrived. Would a specific exemption for JSON decoders be sufficient or is this a pattern we can expect to see many places? If we could permit these cases, is there anything to justify permitting the list filter example which crashes?

**Comment by jvoigtlaender** [comment]
> Would a specific exemption for JSON decoders be sufficient or is this a pattern we can expect to see many places?

It is a pattern I expect to see in libraries and client code that the compiler knows nothing special about. It does, for example, appear in current parser libraries published on the package site, and in client code using those libraries. I don't see how you would make the compiler permit these cases, so this question:
> If we could permit these cases, is there anything to justify permitting the list filter example which crashes?

becomes immaterial to me.

**Comment by andersk** [comment]
A conservative rule would be the following: **there should be no cycle of definitions where each name appears free in a strict position in the definition of the next**. A “strict position” is any position that is guaranteed to be evaluated when the definition is evaluated—so this would exempt function bodies, conditional bodies, and the right side of short-circuiting operators.

All valid code satisfies this rule, because Elm is a strict language and it’s not possible to evaluate a cycle of references in strict position.

Some invalid code also satisfies this rule:

```elm
bad1 = (\() -> bad1) ()
bad2 = if True then bad2 else ()
bad3 = True && bad3
```

However, we can’t hope to catch all invalid code without overly severe restrictions. The above discussion mentions [good use cases](https://github.com/elm-lang/elm-compiler/issues/873#issuecomment-220797386) for values with self-references inside function bodies.

I find it hard to imagine reasons to write values with self-references inside conditionals or short-circuiting operators, so it is tempting to disallow those as well. But perhaps someone has use cases for code shaped like this:

```elm
x = if flag then a else f y
y = if flag then g x else b
```

And if the goal is only to discourage accidental foot-shooting, not to prevent intentional foot-shooting, then keeping that allowed that seems like a reasonable tradeoff.

**Comment by andersk** [comment]
Oh—except that my last example doesn’t work in Elm today, because the compiler generates

```javascript
var x = flag ? a : f(y);
var y = flag ? g(x) : b;
```

and `y` is `undefined` when the first line is executed.

So this slightly tightened version of the rule really is necessary: **there should be no cycle of definitions where each name appears outside a function body in the definition of the next**.

**Comment by andersk** [comment]
Additionally, Elm would need to take this rule into account when sorting definitions, which it doesn’t today:

```elm
a = f True
f b = if b then "Hello, world!" else a
```

should totally work, but currently results in a call to `f` before `f` is defined.

**Comment by mgold** [comment]
Evan is aware of the "self-references must appear inside a lambda/function body" rule. Elegantly, recursive functions are not a special case, they just use a lambda as the root of their defintion, so any self-reference is inside of it.

**Comment by markhamburg** [comment]
It's a long time since I've done any compiler work, but I believe that the following structure would address this. Values can be viewed as going through three phases: allocation, initialization, and use. Allocation is always safe because it just reserves storage. Initialization amounts to filling in the structure. Use involves reading values back out of the structure. Once allocated, it is safe to assign references to the value elsewhere as part of initialization, but we can't safely use a value in any other way until it has been fully initialized and anything it references has been fully initialized. A reasonable — though pessimistic assumption is that a use on any conditional branch will happen and that a parameter passed to a function will be used. Type constructors and lambda constructors could be invoked with values that were allocated but not initialized because the compiler can readily determine that no use will be made of the actual contents of the values.

In any event, I think Elm really only needs to allow the mutually recursive cases that are deemed useful in practice — e.g., mutually recursive functions — even if it can be argued that a case like that cited by andersk above could meaningfully work — i.e., the compiler should be free to say "defining a calls f which needs a to be defined".

**Comment by alex-lew** [comment]
Re: Evan's original post, is the desired behavior of `let ones = 1 :: ones` eventually to shadow the outer `ones` with a longer-by-one `ones` accessible in the let body? It would be nice to have this (and in general, easier-to-reason-about scoping behavior in let expressions, names bound during pattern matching, etc.).

**Comment by mgold** [comment]
Evan showed me how this works in OCaml, and the attitude I got was "neat trick, but it's not useful and it's confusing". I tend to agree - it's an infinite list of the number one, or it's a tree where, wherever you go, you're in the same place. Isn't there some way to do that without a language feature? Also, Elm doesn't have infinite or self-looping data structures at the moment (immutability ensures no reference cycles) so this would "feel" very un-elm-y.

**Comment by alex-lew** [comment]
@mgold Thank you! Totally agree. I guess I was thinking of the case where `ones` is already defined in the outer scope:

```
ones = [1, 1, 1]
let ones = 1 :: ones in ones
```

Would this evaluate to `[1, 1, 1, 1]` in the ideal Elm, or be an error? My thinking is that since you can usually access the other scope (`let newOnes = 1 :: ones`) and you can usually shadow variables in the outer scope (`let ones = somethingTotallyDifferent`), you should be able to do both at once (`let ones = 1 :: ones`). But I could get behind the "it's an error" case too.

**Comment by mgold** [comment]
Okay, that's not completely insane. But the definition looks like it's recursive when it's not, so I think it's better to call it an error. I get the "locally modified copy of some higher-scope thing" problem, but can't you do:

```elm
ones = [1, 1, 1]
ones' = 1 :: ones
```

**Comment by pdamoc** [comment]
I was going to open a new issue but maybe this pattern belongs here: 

```elm
import Html exposing (text)

( val1, val2 ) =
  ( 80, val1 + 80  )

main =
  text <| toString (val1, val2)
```
No compiler error; outputs `(80, NaN)`

**Comment by evancz** [comment]
This should be fixed in 0.18

**Comment by evancz** [comment]
Okay, writing something like `x = x + 1` will now result in this message:

![mutation](https://cloud.githubusercontent.com/assets/1658058/19910332/8167810a-a049-11e6-83da-91f4f43d250a.png)
